lib/collection/src/shards/replica_set/update.rs

```diff
@@ -322,91 +322,102 @@ impl ShardReplicaSet {
-                    replica_state.wait_for(
-                        |state| {
-                            peer_ids.iter().all(|peer_id| {
-                                state
-                                    .peers
-                                    .get(peer_id)
-                                    .map(|state| state != &ReplicaState::Active)
-                                    .unwrap_or(true) // not found means that peer is dead
-                            })
-                        },
-                        DEFAULT_SHARD_DEACTIVATION_TIMEOUT,
-                    )
-                })
-                .await?;
+        if !failures.is_empty() {
+            // If there aren't enough successes, report error to user
+            if successes.len() < minimal_success_count {
+                // TODO(resharding): reconsider how we count/deactivate resharding replicas.
+                self.handle_failed_replicas(
+                    failures
+                        .iter()
+                        .filter(|(peer_id, _)| self.peer_is_resharding(*peer_id)),
+                    &self.replica_state.read(),
+                    update_only_existing,
+                );
 
-                if !shards_disabled {
-                    return Err(CollectionError::service_error(format!(
-                        "Some replica of shard {} failed to apply operation and deactivation \
-                         timed out after {} seconds. Consistency of this update is not guaranteed. Please retry. {failure_error}",
-                        self.shard_id, timeout.as_secs()
-                    )));
-                }
+                let (_peer_id, err) = failures.into_iter().next().unwrap();
+                return Err(err);
             }
-        }
 
-        if !failures.is_empty() && successes.len() < minimal_success_count {
-            self.handle_failed_replicas(
-                failures
-                    .iter()
-                    .filter(|(peer_id, _)| self.peer_is_resharding(*peer_id)),
-                &self.replica_state.read(),
-                update_only_existing,
-            );
+            // If there are enough successes, deactivate failed replicas
+            // Failed replicas will automatically recover from another replica ensuring consistency
+            if successes.len() >= minimal_success_count {
+                let wait_for_deactivation = self.handle_failed_replicas(
+                    &failures,
+                    &self.replica_state.read(),
+                    update_only_existing,
+                );
 
-            // completely failed - report error to user
-            let (_peer_id, err) = failures.into_iter().next().expect("failures is not empty");
-            return Err(err);
+                // Wait for replica failures to be accepted, otherwise return consistency error
+                if wait && wait_for_deactivation && !failures.is_empty() {
```

Comment:
```suggestion
                if wait && wait_for_deactivation {
```

`!failures.is_empty()` is redundant here, because we check for it in the outer block